Recommend feature#27

nahokyun opened the pull request:
> ## recommend 기능 구현
- 영화 세부정보를 다룰 MovieController에서 리뷰의 추천 관련 입력을 받아 처리하도록 구현하였습니다.
- 추천 입력의 비즈니스로직을 다룰 RecommendService를 추가하였습니다.
### MovieController.java
```java
@RestController
@RequiredArgsConstructor
@RequestMapping("/movies")
public class MovieController {
    private final RecommendService recommendService;

    @PostMapping("/recommend-reviews")
    @ResponseStatus(HttpStatus.OK)
    public String recommendUpdate(@RequestBody Recommend recommend) {
        recommendService.recommendHandle(recommend);
        return SUCCESS_RECOMMEND_UPDATE.getMessage();
    }
}
```
### RecommendService.java
- recommand와 관련된 입력은 한 url을 통해 들어올것으로 가정하여 RecommendService의 recommendHandler를 통해 등록, 변경, 삭제를 진행하도록 구현하였습니다.
- reviewId,  recommendState의 내용이 들어있는 recommend 객체와 세션에 저장되어있는 memberId를 이용하여 , 추천여부 및 recommendState값에 따라 비교한 후 알맞은 method를 실행하도록 구현하였습니다.
```java
@Service
@RequiredArgsConstructor
public class RecommendService {

    private final RecommendMapper recommendMapper;
    private final HttpSession httpSession;

    public void recommendHandle(Recommend inputRecommend){

        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);
        Long reviewId=inputRecommend.getReviewId();

        if(recommendMapper.findByReviewIdAndMemberId(memberId,reviewId).isPresent()){
            recommendUpdate(inputRecommend, memberId, reviewId);
        }
        else {
            reviewJoin(inputRecommend, memberId);
        }
    }

    private void recommendUpdate(Recommend inputRecommend, Long memberId, Long reviewId) {
        Recommend saved=(recommendMapper.findByReviewIdAndMemberId(memberId, reviewId)).get();

        if(saved.getRecommendState()!= inputRecommend.getRecommendState()){
            recommendMapper.reviewUpdate(inputRecommend);
        }
        else{
            recommendMapper.deleteRecommend(memberId, reviewId);
        }
    }

    private void reviewJoin(Recommend inputRecommend, Long memberId) {
        Recommend recommend = Recommend.builder()
                .memberId(memberId)
                .reviewId(inputRecommend.getReviewId())
                .recommendState(inputRecommend.getRecommendState())
                .build();

        recommendMapper.reviewJoin(recommend);
    }
}
```
### RecommendMapper.xml
- 입력, reviewId와 memberId로 객체탐색, 추천 및 비추천 상태 변경, 같은 상태값 입력시 삭제 등을 위한 쿼리문입니다.
```xml
   <insert id="reviewJoin" useGeneratedKeys="true" keyProperty="recommendId">
        insert into recommend(review_id, member_id, recommend_state, updated_at, created_at)
        values (#{reviewId}, #{memberId}, #{recommendState}, current_timestamp, current_timestamp)
    </insert>

    <select id="findByReviewIdAndMemberId" resultType="Recommend">
        select id, member_id, review_id, recommend_state
        from recommend
        where member_id = #{memberId}
          and review_id = #{reviewId}
    </select>

    <update id="reviewUpdate" parameterType="Recommend">
        update recommend
        set recommend_state=#{recommendState}
        where member_id = #{memberId}
          and review_id = #{reviewId}
    </update>

    <delete id="deleteRecommend">
        delete
        from recommend
        where member_id = #{memberId}
          and review_id = #{reviewId}
    </delete>
```

nahokyun committed:
Poster-regist#22

MainController에서 poster 이미지 등록을 저장하는 posterJoin이 추가되었습니다.

poster 등록을 위해 새로운 poster관련 쿼리를 추가하였습니다.

nahokyun committed:
Poster-regist#22

addPoster에서 파일입출력에 해당하는 부분을 분리하여 saveFile이라는 메소드를 추가하였습니다.

nahokyun committed:
Poster-regist#22

파일 입출력 부분을 추상화하여 FileIO 인터페이스를 만들었고, 현재는 FileIOLocal을 적용하고 추후 S3를 적용하면 쉽게 바꿀수 있도록 설계하였습니다.

PosterMapper부분을 추가하여 MovieMapper와 분리하였습니다.

nahokyun committed:
Poster-regist#22

범용적인 메소드이름으로 변경하였습니다.

한정적이게 사용될 수 있는 예외처리를 상위레벨로 처리를 미뤄 유연한 사용을 할수있도록 변경하였습니다.

properties를 사용하지않는 생성자를 롬복으로 대체하였습니다.

nahokyun committed:
Poster-regist#22

예외가 발생할 부분만 try-catch문에 남기고 나머지 부분은 이후 처리하는것으로 변경하였습니다.

nahokyun committed:
Merge branch 'Movie-regist#18' into Poster-regist#22

nahokyun committed:
recommend 기능 구현#22

추천 기능을 구현하였습니다.

추천관련 입력은 한 url을 통해 들어올것으로 가정하여 RecommendService의 recommendHandle을 통해 등록, 변경, 삭제를 진행하도록 구현하였습니다.

sunwootest commented:
타겟브랜치가 잘못 설정되었네요

sunwootest commented on src/main/java/study/movieservice/service/RecommendService.java:
```diff
@@ -0,0 +1,51 @@
+        if(recommendMapper.findByReviewIdAndMemberId(memberId,reviewId).isPresent()){
+            recommendUpdate(inputRecommend, memberId, reviewId);
+        }
+        else {
+            reviewJoin(inputRecommend, memberId);
+        }
```
else는 생략해도 됩니다

sunwootest commented on src/main/java/study/movieservice/repository/RecommendMapper.java:
```diff
@@ -0,0 +1,17 @@
+public interface RecommendMapper {
+    void reviewJoin(Recommend recommend);
+
+    Optional<Recommend> findByReviewIdAndMemberId(Long memberId, Long reviewId);
```
메소드명과 파라미터 순서가 안맞네요

nahokyun replied:
수정하도록 하겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/RecommendService.java:
```diff
@@ -0,0 +1,51 @@
+    private final RecommendMapper recommendMapper;
+    private final HttpSession httpSession;
+
+    public void recommendHandle(Recommend inputRecommend){
+
+        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);
+        Long reviewId=inputRecommend.getReviewId();
+
+        if(recommendMapper.findByReviewIdAndMemberId(memberId,reviewId).isPresent()){
+            recommendUpdate(inputRecommend, memberId, reviewId);
+        }
+        else {
+            reviewJoin(inputRecommend, memberId);
+        }
+    }
```
recommendHandle이라는 메소드에서 모두 처리하는 것은 바람직하지 않아보입니다.

nahokyun replied:
각각 http 메소드 별로 분리해보도록 하겠습니다.

sunwootest reviewed (commented):
같은 url이라도 http 메소드에 따라 서로 다른 방식으로 통신할 수 있습니다. 굳이 하나의 서비스에 CRUD를 담을필요는 없어보입니다 :)

nahokyun committed:
recommend 기능 구현#22

http 메소드별로 MovieContoller에서 다르게 매핑해주어 RecommendService에서 처리하도록 분리하였습니다.

nahokyun commented:
> ### 변경된점
- http 메소드별로 다른 비즈니스 로직을 실행하도록 수정하였습니다.
- recommendHandle을 제거하고 각각 recommendJoin, recommendUpdate, recommendDelete 메소드로 분리하였습니다.
- SessionManager.java에 getMemberId메소드를 추가하여 RecommendService에서 HttpSession에 의존하는것을 해결하였습니다.
## MainController.java
### 변경 전 
```java
@PostMapping("/recommend-reviews")
    @ResponseStatus(HttpStatus.OK)
    public String recommendUpdate(@RequestBody Recommend recommend) {
        recommendService.recommendHandle(recommend);
        return SUCCESS_RECOMMEND_UPDATE.getMessage();
    }
```
### 변경 후 
```java
@PostMapping("/recommend-reviews")
    @ResponseStatus(HttpStatus.OK)
    public String recommendJoin(@RequestBody Recommend recommend) {
        recommendService.recommendJoin(recommend);
        return SUCCESS_RECOMMEND_JOIN.getMessage();
    }

    @PatchMapping("/recommend-reviews")
    @ResponseStatus(HttpStatus.OK)
    public String recommendUpdate(@RequestBody Recommend recommend) {
        recommendService.recommendUpdate(recommend);
        return SUCCESS_RECOMMEND_UPDATE.getMessage();
    }

    @DeleteMapping("/recommend-reviews")
    @ResponseStatus(HttpStatus.OK)
    public String recommendDelete(@RequestBody Recommend recommend) {
        recommendService.recommendDelete(recommend);
        return SUCCESS_RECOMMEND_DELETE.getMessage();
    }
```
## RecommendService.java
### 변경 전 
```java
public void recommendHandle(Recommend inputRecommend){

        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);
        Long reviewId=inputRecommend.getReviewId();

        if(recommendMapper.findByReviewIdAndMemberId(memberId,reviewId).isPresent()){
            recommendUpdate(inputRecommend, memberId, reviewId);
        }
        else {
            reviewJoin(inputRecommend, memberId);
        }
    }
```
### 변경 후 
```java
    public void recommendJoin(Recommend inputRecommend) {
        if (recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), inputRecommend.getMemberId()).isEmpty()) {

             Long memberId = sessionManager.getMemberId();

            Recommend recommend = Recommend.builder()
                    .memberId(memberId)
                    .reviewId(inputRecommend.getReviewId())
                    .recommendState(inputRecommend.getRecommendState())
                    .build();

            recommendMapper.reviewJoin(recommend);
        }
        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
    }

    public void recommendUpdate(Recommend inputRecommend) {
        Long memberId = sessionManager.getMemberId();

        if (recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), memberId).isPresent())
            recommendMapper.reviewUpdate(inputRecommend, memberId);
        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
    }

    public void recommendDelete(Recommend inputRecommend) {
        Long memberId = sessionManager.getMemberId();

        if (recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), memberId).isPresent())
            recommendMapper.deleteRecommend(memberId, inputRecommend.getReviewId());
        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
    }
```

Jeonminje commented on src/main/java/study/movieservice/service/RecommendService.java:
```diff
@@ -0,0 +1,51 @@
+    private final RecommendMapper recommendMapper;
+    private final SessionManager sessionManager;
+
+    public void recommendJoin(Recommend inputRecommend) {
+        if (recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), inputRecommend.getMemberId()).isEmpty()) {
+
+            Long memberId = sessionManager.getMemberId();
+
+            Recommend recommend = Recommend.builder()
+                    .memberId(memberId)
+                    .reviewId(inputRecommend.getReviewId())
+                    .recommendState(inputRecommend.getRecommendState())
+                    .build();
+
+            recommendMapper.reviewJoin(recommend);
+        }
+        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
+    }
```
혹시 상기 로직에 대해 설명 해 주실수 있으실까요??
그리고

```java
if (recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), inputRecommend.getMemberId()).isEmpty())
```
이 부분이 제가 이해한 것으로는 테이블에 데이터가 있는지 확인하는 용도로 쓰이는 것 같은데, 제가 이해한 바가 맞다면 exist를 활용하는 쿼리를 짜는 것이 좋아보입니다!

nahokyun replied:
우선 recommendJoin의 로직은 inputRecommend의 reviewId와 session에있는 memberId를 통해 데이터 존재여부를 조회 후 존재하지 않을때 recommend 테이블에 추가하는 로직입니다. 
다만 저부분은 지금 확인해보니 테스트용으로 session이 아닌 inputRecommend에 reviewId, recommendState뿐만 아니라 memberId까지 입력을 넣고 테스트하다 보니 적용후 제대로된 수정을 하지못했네요 바로 수정하겠습니다!
또한 if문 같은 경우는 말씀해주신대로 recommend의 정보를 존재여부만 사용하는것이므로 exist를 활용하는것이 더 좋아보이네요 수정해보겠습니다.

Jeonminje commented on src/main/java/study/movieservice/repository/RecommendMapper.java:
```diff
@@ -0,0 +1,18 @@
+
+@Mapper
+public interface RecommendMapper {
+    void reviewJoin(Recommend recommend);
```
쿼리를 보니 review 테이블과 관계가 없는거 같은데, review가 붙어있어서 헷갈릴 거 같아요..!
그리고 join 보다 save가 어떠실까요?? save가 좀 더 와닿을 거 같아서요!!

nahokyun replied:
Service에서는 이름을 수정했는데 mapper부분까지는 미처 수정하지않았네요. 좀더 확실하게 메소드 명을 수정하도록 하겠습니다!

nahokyun committed:
recommend 기능 구현#22

recommendMapper의 메소드명을 직관적으로 알수있도록 변경하였습니다.

findByReviewIdAndMemberId의 쿼리를 exists를 통하여 불필요하게 객체를 가져오는것을 boolean타입으로 받도록 변경하였습니다.

메소드별 javadoc을 추가하였습니다.

nahokyun commented:
> ### 변경된점
- recommendMapper의 메소드명을 직관적으로 알수있도록 변경하였습니다.
- exists를 활용하여 findByReviewIdAndMemberId를 목적에 맞는 boolean타입으로 반환하도록 변경하였습니다.
- 메소드별 javadoc을 추가로 작성하였습니다.
## RecommendMapper.xml
### 변경 전 
```xml
    <select id="findByReviewIdAndMemberId" resultType="Recommend">
        select id, member_id, review_id, recommend_state
        from recommend
        where member_id = #{memberId}
          and review_id = #{reviewId}
    </select>
```
### 변경 후 
```xml
    <select id="findByReviewIdAndMemberId" resultType="boolean">
        select exists(
            select review_id
            from recommend
            where member_id = #{memberId}
            and review_id = #{reviewId}
        )
    </select>
```
## RecommendService.java
### 변경 전 
```java
    public void recommendJoin(Recommend inputRecommend) {
        if (recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), inputRecommend.getMemberId()).isEmpty()) {

            Long memberId = sessionManager.getMemberId();

            Recommend recommend = Recommend.builder()
                    .memberId(memberId)
                    .reviewId(inputRecommend.getReviewId())
                    .recommendState(inputRecommend.getRecommendState())
                    .build();

            recommendMapper.reviewJoin(recommend);
        }
        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
    }

```
### 변경 후 
```java
    /**
     * 추천입력을 받았을때 review 테이블의 해당객체를 등록하는 메소드
     *
     * @param inputRecommend
     * 어떤 리뷰인지 알려주는 reviewId와
     * 1일때 추천, 0일때 비추천을 나타내는 recommendState 정보보가담겨있는 Recommend 객체
     * @throws IllegalArgumentException
     * 만약 reviewId, memberId가 일치하는 객체가 DB에 없을때 발생
     */
    public void recommendSave(Recommend inputRecommend) {
        Long memberId = sessionManager.getMemberId();

        if (!recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), memberId)) {
            Recommend recommend = Recommend.builder()
                    .memberId(memberId)
                    .reviewId(inputRecommend.getReviewId())
                    .recommendState(inputRecommend.getRecommendState())
                    .build();

            recommendMapper.recommendSave(recommend);
        }
        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
    }

```

sunwootest commented on src/main/java/study/movieservice/service/RecommendService.java:
```diff
@@ -0,0 +1,74 @@
+    public void recommendSave(Recommend inputRecommend) {
+        Long memberId = sessionManager.getMemberId();
+
+        if (!recommendMapper.findByReviewIdAndMemberId(inputRecommend.getReviewId(), memberId)) {
+            Recommend recommend = Recommend.builder()
+                    .memberId(memberId)
+                    .reviewId(inputRecommend.getReviewId())
+                    .recommendState(inputRecommend.getRecommendState())
+                    .build();
+
+            recommendMapper.recommendSave(recommend);
+        }
+        else throw new IllegalArgumentException(FAILED_RECOMMEND_REQUEST.getMessage());
```
코딩 스타일에 대해서 피드백드리겠습니다. if문 내부에 예외상황을 throw하게 구현하면 코드가 더 간결해집니다. 그렇게 되면 굳이 else문도 작성할 필요가 없어지겠죠.

nahokyun committed:
recommend 기능 구현#22

if문 내부에 예외상황을 두어 불필요한 else문을 제거하였습니다.

nahokyun commented:
if문 내부에 예외상황을 두어 불필요한 else문을 제거하였습니다.
